File: src/server/keymap.ts
/**
 * Maps browser `KeyboardEvent.key` values to robotjs key names.
 *
 * robotjs can only toggle a fixed set of named keys plus ASCII letters and
 * digits. Printable characters outside that set (punctuation, accented
 * letters) are injected as literal text on key-down instead, which types the
 * character but cannot participate in held-key shortcuts — an acceptable
 * trade-off documented in the README.
 */

export type MappedKey = { type: 'toggle'; key: string } | { type: 'text'; text: string };

const SPECIAL_KEYS: Record<string, string> = {
  ' ': 'space',
  Alt: 'alt',
  ArrowDown: 'down',
  ArrowLeft: 'left',
  ArrowRight: 'right',
  ArrowUp: 'up',
  Backspace: 'backspace',
  Control: 'control',
  Delete: 'delete',
  End: 'end',
  Enter: 'enter',
  Escape: 'escape',
  Home: 'home',
  Insert: 'insert',
  Meta: 'command',
  PageDown: 'pagedown',
  PageUp: 'pageup',
  Shift: 'shift',
  Tab: 'tab',
};

for (let i = 1; i <= 24; i++) {
  SPECIAL_KEYS[`F${i}`] = `f${i}`;
}

/** Map a `KeyboardEvent.key` to an injectable action, or null to drop it. */
export function mapKey(key: string): MappedKey | null {
  // Own-property check only: a bare `SPECIAL_KEYS[key]` would also resolve
  // inherited members like "constructor" or "toString" to functions, which
  // would then be handed to the native injector and throw.
  const special = Object.hasOwn(SPECIAL_KEYS, key) ? SPECIAL_KEYS[key] : undefined;
  if (special) return { type: 'toggle', key: special };
  if (key.length !== 1) return null;
  if (/^[a-zA-Z0-9]$/.test(key)) return { type: 'toggle', key: key.toLowerCase() };
  // Any other single printable character (punctuation, accented letters, …).
  return { type: 'text', text: key };
}

File: desktop/src-tauri/src/keymap.rs
//! Maps browser `KeyboardEvent.key` values to `enigo` keys.
//!
//! Named keys map to dedicated `enigo::Key` variants; ASCII letters/digits map
//! to `Key::Unicode`; any other single printable character is injected as text
//! on key-down (mirroring the Node server's behavior).

use enigo::Key;

pub enum MappedKey {
    Toggle(Key),
    Text(String),
}

pub fn map_key(key: &str) -> Option<MappedKey> {
    if let Some(named) = named_key(key) {
        return Some(MappedKey::Toggle(named));
    }
    let mut chars = key.chars();
    let (Some(ch), None) = (chars.next(), chars.next()) else {
        return None; // multi-char name we don't recognize
    };
    if ch.is_ascii_alphanumeric() {
        Some(MappedKey::Toggle(Key::Unicode(ch.to_ascii_lowercase())))
    } else {
        Some(MappedKey::Text(ch.to_string()))
    }
}

fn named_key(key: &str) -> Option<Key> {
    let k = match key {
        " " => Key::Space,
        "Alt" => Key::Alt,
        "ArrowDown" => Key::DownArrow,
        "ArrowLeft" => Key::LeftArrow,
        "ArrowRight" => Key::RightArrow,
        "ArrowUp" => Key::UpArrow,
        "Backspace" => Key::Backspace,
        "Control" => Key::Control,
        "Delete" => Key::Delete,
        "End" => Key::End,
        "Enter" => Key::Return,
        "Escape" => Key::Escape,
        "Home" => Key::Home,
        "Insert" => Key::Insert,
        "Meta" => Key::Meta,
        "PageDown" => Key::PageDown,
        "PageUp" => Key::PageUp,
        "Shift" => Key::Shift,
        "Tab" => Key::Tab,
        "F1" => Key::F1,
        "F2" => Key::F2,
        "F3" => Key::F3,
        "F4" => Key::F4,
        "F5" => Key::F5,
        "F6" => Key::F6,
        "F7" => Key::F7,
        "F8" => Key::F8,
        "F9" => Key::F9,
        "F10" => Key::F10,
        "F11" => Key::F11,
        "F12" => Key::F12,
        _ => return None,
    };
    Some(k)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn maps_named_and_alnum_keys() {
        assert!(matches!(
            map_key("Enter"),
            Some(MappedKey::Toggle(Key::Return))
        ));
        assert!(matches!(map_key(" "), Some(MappedKey::Toggle(Key::Space))));
        assert!(matches!(
            map_key("A"),
            Some(MappedKey::Toggle(Key::Unicode('a')))
        ));
    }

    #[test]
    fn routes_punctuation_to_text() {
        assert!(matches!(map_key("."), Some(MappedKey::Text(_))));
        assert!(matches!(map_key("é"), Some(MappedKey::Text(_))));
    }

    #[test]
    fn drops_uninjectable_keys() {
        assert!(map_key("CapsLock").is_none());
        assert!(map_key("Unidentified").is_none());
    }
}
